hphp/runtime/vm/jit/vasm-lower.cpp

```diff
@@ -33,6 +33,55 @@ namespace HPHP { namespace jit {
 
 ///////////////////////////////////////////////////////////////////////////////
 
+namespace detail {
+
+jit::vector<int> computeDominatingConstraints(Vunit& unit) {
+  jit::vector<int> vregConstraints(unit.blocks.size(), -1);
+
+  // Build reverse postorder list of blocks
+  auto const rpo = sortBlocks(unit);
+
+  // Calculate each block's dominating vreg constraint. The lowerers will use
+  // this as the initial vreg constraint for each block, adjusting as they
+  // encounter vregallow/vregrestrict during lowering.
+  vregConstraints[rpo[0]] = 0;
+  for (auto b : rpo) {
+    // Iterate through the instructions updating the constraint accordingly.
+    auto con = vregConstraints[b];
+    for (size_t i = 0; i < unit.blocks[b].code.size(); ++i) {
+      auto& inst = unit.blocks[b].code[i];
+      if (inst.op == Vinstr::vregrestrict) {
+        con++;
+      } else if (inst.op == Vinstr::vregallow) {
+        con--;
+      }
+    }
+    assert_flog(con >= 0,
+		"Dominating constraint can't be unknown.");
```

Comment:
If we're going to allow "nesting" of constraints, then we probably don't want this "unknown" behavior.  In particular, if we're going to have `vregallow{}` and `vregrestrict{}` be intrinsic to PhysRegSaver, then we need to let the `vregallow{}` in the constructor change the level to -1, then have the `vregrestrict{}` change the level back to 0, so that PhysRegSaver will still work without being dominated by a `vregrestrict{}` (or more generally, without being in a context where the constraint level is `>= 1`).

The check for "can I make new Vregs" will then be `vreg_level <= 0`.

Also, you commented:

> Note: I implemented this as suggested, but it drops the nesting concept a little. I use the integer to indicate the nested depth of the vregrestricts when calculating the vregConstraints. But then when using this env.allow_vreg bool, we lose the nest depth. In a scenario where vregConstraints[b] == 3, a vregallow within a block will allow new vregs.
>
> Is that you intended behavior? I think it needs to be.

Okay, so... why not change `allow_vreg` to be an int (`vreg_level` or such), and have `vregallow{}` and `vregrestrict{}` decrement and increment it as usual?